realm/realm-library/src/androidTest/java/io/realm/BulkInsertTests.java

```diff
@@ -317,7 +319,33 @@ public void insertToRealm_NullPrimaryKey() {
     }
 
     @Test
-    public void insertOrUpdateToRealm() {
+    public void insert_duplicatedPrimaryKeyFails() {
+
+        // Single object with 2 references to two objects with the same ID
+        AllJavaTypes obj = new AllJavaTypes(2);
+        obj.setFieldList(new RealmList<AllJavaTypes>(new AllJavaTypes(1), new AllJavaTypes(1)));
+        realm.beginTransaction();
+        try {
+            realm.insert(obj);
+            fail();
+        } catch (RealmPrimaryKeyConstraintException ignored) {
```

Comment:
Because you cannot `insert` two objects with the same PK. In that case you should have called `insertOrUpdate`. We cannot really just say it is undefined behaviour as it would be a serious data corruption if you suddenly had two objects with the same primary key.